bpf/lib/lb.h

```diff
@@ -1748,8 +1752,341 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	 */
+	csum = 0;
+
+	#define body(i) if ((i) > sample_len) \
+			goto csumend; \
+		if (data + sizeof(struct ethhdr) + (i) + sizeof(__u16) > data_end) \
+			goto csumend; \
+		csum += *(__u16 *)(data + sizeof(struct ethhdr) + (i));
+
+	#define body4(i) body(i)\
+		body(i + 2) \
+		body(i + 4) \
+		body(i + 6)
+
+	#define body16(i) body4(i)\
+		body4(i + 8) \
+		body4(i + 16) \
+		body4(i + 24)
+
+	#define body128(i) body16(i)\
+		body16(i + 32) \
+		body16(i + 64) \
+		body16(i + 96)
+
+	body128(0)
+	body128(256)
+	body128(512)
+	body128(768)
+	body128(1024)
+
+csumend :
+
+	/* We need to insert a IPv4 and ICMP header before the original packet.
+	 * Make that room.
+	 */
+
+#if __ctx_is == __ctx_xdp
+	xdp_adjust_head(ctx, 0 - (int)(sizeof(struct ipv6hdr) + sizeof(struct icmp6hdr)));
+#else
+	skb_adjust_room(ctx, sizeof(struct ipv6hdr) + sizeof(struct icmp6hdr), BPF_ADJ_ROOM_MAC, 0);
+#endif
+
+	/* changing size invalidates pointers, so we need to re-fetch them. */
+	data = ctx_data(ctx);
+	data_end = ctx_data_end(ctx);
+
+	/* Bound check all 3 headers at once. */
+	if (data + inner_offset > data_end)
+		return DROP_INVALID;
+
+	/* Write reversed eth header, ready for egress */
+	ethhdr = data;
+	memcpy(ethhdr->h_dest, smac.addr, sizeof(smac.addr));
+	memcpy(ethhdr->h_source, dmac.addr, sizeof(dmac.addr));
+	ethhdr->h_proto = bpf_htons(ETH_P_IPV6);
+
+	/* Write reversed ip header, ready for egress */
+	ip6 = data + sizeof(struct ethhdr);
+	ip6->version = 6;
+	ip6->priority = 0;
+	ip6->flow_lbl[0] = 0;
+	ip6->flow_lbl[1] = 0;
+	ip6->flow_lbl[2] = 0;
+	ip6->payload_len = bpf_htons(sizeof(struct icmp6hdr) + (__u16)sample_len);
+	ip6->nexthdr = IPPROTO_ICMPV6;
+	ip6->hop_limit = 64;
+	memcpy(&ip6->daddr, &saddr, sizeof(struct in6_addr));
+	memcpy(&ip6->saddr, &daddr, sizeof(struct in6_addr));
+
+	/* Write reversed icmp header */
+	icmphdr = data + sizeof(struct ethhdr) + sizeof(struct ipv6hdr);
+	icmphdr->icmp6_type = ICMPV6_DEST_UNREACH;
+	icmphdr->icmp6_code = ICMPV6_ADDR_UNREACH;
+	icmphdr->icmp6_cksum = 0;
+	icmphdr->icmp6_dataun.un_data32[0] = 0;
+
+	/* Add the ICMP header to the checksum (only type and code are non-zero) */
+	csum += ((__u16)icmphdr->icmp6_code) << 8 | (__u16)icmphdr->icmp6_type;
+
+	/* Fill pseudo header */
+	memcpy(&pseudo_header.fields.src_ip, &ip6->saddr, sizeof(struct in6_addr));
+	memcpy(&pseudo_header.fields.dst_ip, &ip6->daddr, sizeof(struct in6_addr));
+	pseudo_header.fields.top_level_length = bpf_htonl(sizeof(struct icmp6hdr) +
+						(__u32)sample_len);
+	__bpf_memzero(pseudo_header.fields.zero, sizeof(pseudo_header.fields.zero));
+	pseudo_header.fields.next_header = IPPROTO_ICMPV6;
+
+	#pragma unroll
+	for (i = 0; i < 20; i++)
+		csum += pseudo_header.words[i];
+
+	icmphdr->icmp6_cksum = csum_fold(csum);
+
+	/* Redirect ICMP to the interface we received it on. */
+	cilium_dbg_capture(ctx, DBG_CAPTURE_DELIVERY,
+			   ctx_get_ifindex(ctx));
+	return ctx_redirect(ctx, ctx_get_ifindex(ctx), 0);
```

Comment:
Well we currently don't since I glossed over that part of the spec :fearful:. Any suggestions on how to add that? 

First instinct is to do so in BPF, add a map to track messages and just drop the incoming packet if we hit the quota. Though I suspect someone will insist on making such a mechanism generic and reusable to warrant the cost of introducing a new map.

Second thought is to re-use parts of the bandwidth manager. Which would mean that if we want this behavior to be the default (as it should be for k8s conformance) that bandwidth manager is always enabled. We might be able to just re-use the TC setup in the background so we can use it to rate-limit, without having to enable the whole feature. But it sounds ticky in any case.

I would urge a follow up PR for this works since it seems significant to add, even though that means we technically are not RFC compliant for now. I would argue that its of less importance in this code path anyway since we specifically deal with replies for services without backends, which is a very specific scenario as apposed to something like a echo reply.

Reply:
I'm not quite sure about the "map to track messages" approach, the standard just suggests a token bucket implementation. Bandwidth manager seems like it may have more in common.

As a general principle, I would be more comfortable with merging such functionality into a branch and potentially breaking up individual tasks into PRs to target that branch, then only merge the branch into main when the feature is complete / standards-compliant. I worry that it's too easy to say "well that's a big item so we'll do it later" and then later becomes ~forever, and Cilium ends up releasing a version that violates well-established standards. This is part of the spec for the feature that's being added in this PR.

Reply:
> I worry that it's too easy to say "well that's a big item so we'll do it later" 

I personally don't mind putting it all into one PR, but I have been told that smaller PRs are easier to review hence the initial suggestion. :grin: 

> I would be more comfortable with merging such functionality into a branch and potentially breaking up individual tasks into PRs to target that branch

If we can merge this into a `feature/...` branch with reviewed PRs and then merge the whole branch into main as a whole that would be optimal in my opinion.

> I'm not quite sure about the "map to track messages" approach, the standard just suggests a token bucket implementation. Bandwidth manager seems like it may have more in common.

I meant implementing something like a token bucket algorithm with a map as storage. But I agree that reusing the bandwidth manager would be better than making something new at this point.

> [RFC4443](https://datatracker.ietf.org/doc/html/rfc4443) 2.4.f requires implementations to limit how many ICMPv6 messages

Motivation behind this PR is reports of users hitting this issue. Would it be acceptable to rip out the IPv6 path since its only the ICMPv6 spec that requires rate limiting? That way we can ship something in a reasonable amount of time, given I expect the bandwidth manager route to take significant time

Reply:
I've created `feature/main/svc-icmp-response`, you should be able to submit PRs against that branch. The goal should be to ensure that merges into that branch match the expectations for merging into `main`. Periodically if necessary we can rebase it (I think you should have the ability to do this, but let me know if not). Once we think the feature is complete we can submit a PR to pull that branch into main.

> Motivation behind this PR is reports of users hitting this issue. Would it be acceptable to rip out the IPv6 path since its only the ICMPv6 spec that requires rate limiting? That way we can ship something in a reasonable amount of time, given I expect the bandwidth manager route to take significant time

This has been the behaviour for _years_. This is a brand new feature being submitted half way through the v1.15 development cycle. I don't see the urgency. I'll note that arguments like this are a common cause of tech debt and also a big reason why various features don't work with IPv6. By all means we can have a discussion with the relevant sig-datapath folks to set the expectations according to the shared community quality bar, but I don't really buy the argument why corners must be cut here.